xnmt/self_attention.py

```diff
@@ -0,0 +1,87 @@
+import numpy as np
+import dynet as dy
+from math import sqrt
+
+from xnmt.events import register_xnmt_handler, handle_xnmt_event
+from xnmt.expression_sequence import ExpressionSequence, ReversedExpressionSequence
+from xnmt.param_collection import ParamManager
+from xnmt.param_init import GlorotInitializer, ZeroInitializer
+from xnmt.persistence import serializable_init, Serializable, bare, Ref
+from xnmt.transducer import SeqTransducer
+
+class MultiHeadAttentionSeqTransducer(SeqTransducer, Serializable):
+  """
+  This implements the Multi-headed attention layer of "Attention is All You Need":
+  https://papers.nips.cc/paper/7181-attention-is-all-you-need.pdf
+
+  Args:
+
+  """
+  yaml_tag = '!MultiHeadAttentionSeqTransducer'
+
+  @register_xnmt_handler
+  @serializable_init
+  def __init__(self,
+               input_dim=Ref("exp_global.default_layer_dim"),
+               param_init=Ref("exp_global.param_init", default=bare(GlorotInitializer)),
+               bias_init=Ref("exp_global.bias_init", default=bare(ZeroInitializer)),
+               num_heads=8,
+               yaml_path=None):
+    assert(input_dim % num_heads == 0)
+
+    param_collection = ParamManager.my_params(self)
+
+    self.input_dim = input_dim
+    self.num_heads = num_heads
+    self.head_dim = input_dim // num_heads
+
+    # self.Wq, self.Wk, self.Wv, self.Wo = [param_collection.add_parameters(dim=(input_dim, input_dim), init=param_init) for _ in range(4)]
+    # self.bq, self.bk, self.bv, self.bo = [param_collection.add_parameters(dim=(input_dim), init=bias_init) for _ in range(4)]
+    self.Wq, self.Wk, self.Wv, self.Wo = [param_collection.add_parameters(dim=(input_dim, input_dim)) for _ in range(4)]
+    self.bq, self.bk, self.bv, self.bo = [param_collection.add_parameters(dim=(1, input_dim)) for _ in range(4)]
```

Comment:
These should probably use the passed parameter initializers, similar to e.g. https://github.com/neulab/xnmt/blob/master/xnmt/linear.py#L29

Reply:
Fixed, thanks!